src/main/java/com/paloaltonetworks/panorama/api/methods/PanoramaApiClient.java

```diff
@@ -0,0 +1,575 @@
+        String vmAuthKeyString = vmAuthKeyResponse.getShowResult();
+        if (vmAuthKeyString == null) {
+            throw new NullPointerException("Null result generating vmAuth with params: " + queryStrings);
+        }
+
+        String[] temp = vmAuthKeyString.split(" ");
+
+        if (temp.length < 4) {
+            throw new Exception("Error parsing vmAuth Response: it contains no vmAuth key ");
+        }
+
+        return temp[3];
+    }
+
+    @SuppressWarnings("unchecked")
+    public List<String> showDevices() throws Exception {
+        Map<String, String> queryStrings = makeRequestParams(null, OP_TYPE, null, null, SHOW_DEVICES_ALL_CMD);
+        ShowDeviceResponse showDeviceResponse = getRequest(queryStrings, ShowDeviceResponse.class);
+        List<DeviceEntry> deviceGroups = emptyList();
+
+        if (showDeviceResponse.getShowDeviceResult() != null){
+            deviceGroups = showDeviceResponse.getShowDeviceResult().getShowDeviceEntries();
+            deviceGroups = deviceGroups != null ? deviceGroups : emptyList();
+        }
+
+        LOG.info("Pan Devices List: " + deviceGroups);
+        return deviceGroups.stream().filter(de -> de != null).map(de -> de.getName())
+                            .collect(toList());
+    }
+
+    public List<String> showDeviceGroups() throws Exception {
+        Map<String, String> queryStrings = makeRequestParams(null, OP_TYPE, null, null, SHOW_DEVICEGROUPS_CMD);
+        List<DeviceGroupsEntry> deviceGroups = emptyList();
+        DeviceGroupResponse deviceGroupResponse = getRequest(queryStrings, DeviceGroupResponse.class);
+
+        if (deviceGroupResponse.getDeviceGroups() != null) {
+            deviceGroups = deviceGroupResponse.getDeviceGroups().getEntries();
+            deviceGroups = deviceGroups != null ? deviceGroups : emptyList();
+        }
+
+        List<String> panDeviceGroups = deviceGroups.stream().filter(dg -> dg != null).map(dg -> dg.getName())
+                                                   .collect(toList());
+
+        LOG.info("Pan Device Groups List: " + panDeviceGroups);
+
+        return panDeviceGroups;
+    }
+
+    public String addDeviceGroup(String name, String description) throws Exception {
+        LOG.info("Adding device group " + name);
+        // "<entry name=\'" + name + "\'><description>" + description + "</description><devices/></entry>";
+        String element = makeEntryElement(name, null, description, null);
+        Map<String, String> queryStrings = makeSetConfigRequestParams(XPATH_DEVGROUP_PREFIX, element, null);
+        String status = getRequest(queryStrings, SetConfigResponse.class).getStatus();
+        String errorMessage = String.format(
+                "Commit failed when adding Device Group Name: %s and Description: %s ", name, description);
+        configCommitOrThrow(errorMessage);
+        return status;
+    }
+
+     public String addDAG(String name, List<String> tags) throws Exception {

```

Comment:
keep only relavent api's?

We dont add Dag's anymore right? we only add addresses and tag them with the OSC prefix-sg-id